src/MongoDB.Bson/Serialization/Serializers/GuidSerializer.cs

```diff
@@ -126,16 +155,39 @@ public override void Serialize(BsonSerializationContext context, BsonSerializati
 
             switch (_representation)
             {
+#pragma warning disable 618
                 case BsonType.Binary:
-                    var writerGuidRepresentation = bsonWriter.Settings.GuidRepresentation;
-                    if (writerGuidRepresentation == GuidRepresentation.Unspecified)
+                    var guidRepresentation = _guidRepresentation;
+                    if (BsonDefaults.GuidRepresentationMode == GuidRepresentationMode.V2 && guidRepresentation == GuidRepresentation.Unspecified)
+                    {
+                        guidRepresentation = bsonWriter.Settings.GuidRepresentation;
+                    }
+                    if (guidRepresentation == GuidRepresentation.Unspecified)
                     {
                         throw new BsonSerializationException("GuidSerializer cannot serialize a Guid when GuidRepresentation is Unspecified.");
                     }
-                    var bytes = GuidConverter.ToBytes(value, writerGuidRepresentation);
-                    var subType = (writerGuidRepresentation == GuidRepresentation.Standard) ? BsonBinarySubType.UuidStandard : BsonBinarySubType.UuidLegacy;
-                    bsonWriter.WriteBinaryData(new BsonBinaryData(bytes, subType, writerGuidRepresentation));
+                    var bytes = GuidConverter.ToBytes(value, guidRepresentation);
+                    var subType = GuidConverter.GetSubType(guidRepresentation);
+                    if (BsonDefaults.GuidRepresentationMode == GuidRepresentationMode.V2)
+                    {
+                        var binaryData = new BsonBinaryData(bytes, subType, guidRepresentation);
+                        bsonWriter.PushSettings(s => s.GuidRepresentation = GuidRepresentation.Unspecified);
+                        try
+                        {
+                            bsonWriter.WriteBinaryData(binaryData);
+                        }
+                        finally
+                        {
+                            bsonWriter.PopSettings();
+                        }
```

Comment:
This missing call to PushSettings is what was causing tests to fail for Dmitry when he experimented with ObjectSerializer delegating to GuidSerializer.

What's going on here is that GuidSerializer needs to protect itself from a writer having a different GuidRepresentation. Just like we've done numerous times already the way to suppress a writer's default GuidRepresentation behavior is to set the representation to Unspecified.